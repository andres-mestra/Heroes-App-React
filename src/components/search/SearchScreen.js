import *  as React from 'react'
import { useLocation } from 'react-router-dom';
import queryString from 'query-string'
import { HeroListCard } from '../heroes/HeroListCard';
import { useForm } from '../../hooks/useForm';
import { heroes } from '../../data/heroes';
import { getHeroesByName } from '../../selectors/getHeroesByName';



export const SearchScreen = ({ history }) => {

  const location = useLocation();
  //parsea el q de la url, si no esta es q se le asigna por defecto el valor ''
  const { q = ''} = queryString.parse(location.search);
  const [ values, handleInputChange, reset ] = useForm({
    search: q
  })
  const { search } = values;

  //solo se ejecuta si el q cambia
  const heroesFilter = React.useMemo( () => getHeroesByName(q), [q]);

  const handleSubmit =  ( e ) => {
    e.preventDefault()
    //query selector en la url
    history.push(`?q=${ search }`)
  }

  return (
    <article>
      <h1>Search Screen</h1>
      <hr/>
      <div className="row ">
        <section className="col-lg-5 col-sm-12">
          <h4>Search Form</h4>
          <hr/>
          <form className="form-group" onSubmit={ handleSubmit }>
            <input 
              type="text"
              name="search"
              placeholder="Find your hero"
              className="form-control"
              onChange={handleInputChange}
            />

            <button 
              type="submit"
              className="btn btn-outline-primary mt-2"
            >
              Search...
            </button>
          </form>

        </section>
        <section className="col-sm-12 col-lg-7 ">
          <h4> Results </h4>
          <hr/>
          {
            ( q === '') 
            &&
              <div className="alert alert-info">
                Search a hero
              </div>
          }
          {
            ( q !== '' && heroesFilter.length === 0 ) 
            &&
              <div className="alert alert-danger">
                There is not a hero with { q }
              </div>
          }

          {
            heroesFilter.map( hero => (
              <HeroListCard 
                key={ hero.id }
                { ...hero }
              />
            ))
          }
        </section>
      </div>
    </article>
  )
}
