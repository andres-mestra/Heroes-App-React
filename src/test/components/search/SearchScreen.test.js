import React from 'react'
import { MemoryRouter, Route } from 'react-router-dom'
import { mount } from 'enzyme'
import { SearchScreen } from '../../../components/search/SearchScreen'


describe('Pruebas en <SearchScreen />', () => {
  
  test('debe de mostrar correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search'] }>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter> 
    )
    
    expect( wrapper ).toMatchSnapshot()
    expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a hero')
  })

  test('Debe de mostrar a Batman y el input con el queryString. ', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=batman'] }>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter> 
    )
    
    expect( wrapper.find('HeroListCard').exists() ).toBe(true);
  })
  
  test('Debe de mostrar un error si no se encuetra el hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=0'] }>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter> 
    )
    
    expect( wrapper.find('.alert-danger').text().trim() ).toBe('There is not a hero with 0')
  })
  
  test('debe de llamar el push del history', () => {
    
    const historyMock = {
      push: jest.fn()
    }
     
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=0'] }>
        <Route 
          path="/search" 
          component={ () => <SearchScreen history={historyMock} />} 
        />
      </MemoryRouter> 
    )
    
    wrapper.find('input').simulate('change', {
      target: {
        name:'search',
        value: 'batman'
      }
    })

    wrapper.find('form').prop('onSubmit')({ 
      preventDefault(){}
    })
    
    expect( historyMock.push ).toHaveBeenCalledWith('?q=batman')
  })
  
  
})
