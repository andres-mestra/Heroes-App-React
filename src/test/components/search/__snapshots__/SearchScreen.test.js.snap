// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar correctamente con valores por defecto 1`] = `
<article>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row "
  >
    <section
      className="col-lg-5 col-sm-12"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="form-group"
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </section>
    <section
      className="col-sm-12 col-lg-7 "
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </section>
  </div>
</article>
`;
